core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol2.java

```diff
@@ -54,8 +55,13 @@ protected void run() throws IOException, InterruptedException {
             final String nodeName = request.getProperty("Node-Name");
 
             for (JnlpAgentReceiver recv : JnlpAgentReceiver.all()) {
-                if (recv.handle(nodeName,this))
-                    return;
+                try {
+                    if (recv.handle(nodeName,this))
+                        return;
+                } catch (AbstractMethodError e) {
+                    if (recv.handle(nodeName,new JnlpSlaveHandshake(this)))
```

Comment:
Looks alarming. Why not just retain the original method signature, deprecate it, and delegate?